### PR TITLE
Document continuous integration

### PR DESCRIPTION
**Pull request by baijum**
Reviewers, please see the preview of the documentation here:
https://github.com/baijum/devconsole-operator/blob/ci-readme/openshift-ci/README.md

### CODE REVIEW

**Comment by pratikjagrut** [comment]
@baijum what I think is I see [Openshift CI](https://github.com/openshift/release) reference, but somewhere you should also make direct reference to this file [redhat-developer-devconsole-operator-master.yaml](https://github.com/openshift/release/blob/master/ci-operator/config/redhat-developer/devconsole-operator/redhat-developer-devconsole-operator-master.yaml)  it will help us to understand more.

File: openshift-ci/README.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+### test
+
+The `test` target runs the unit tests.  Some of the test make use mock
+objects. The unit tests doesn't require a dedicated OpenShift cluster
+unlike end-to-end tests.
+
+### e2e
+
+The `e2e` run an end-to-end test against an operator running inside
+the CI cluster pod but connected to a freshly created temperory
```
Comment (Avni-Sharma) [inline openshift-ci/README.md]: slight nit pick- temporary

Comment (baijum) [reply openshift-ci/README.md]: Fixed. 8e6837f5d70386a79c9d8a88f80671d82412ce36

**Comment by baijum** [comment]
> @baijum what I think is I see [Openshift CI](https://github.com/openshift/release) reference, but somewhere you should also make direct reference to this file [redhat-developer-devconsole-operator-master.yaml](https://github.com/openshift/release/blob/master/ci-operator/config/redhat-developer/devconsole-operator/redhat-developer-devconsole-operator-master.yaml) it will help us to understand more.

@pratikjagrut I have added the link in the first paragraph itself.

File: openshift-ci/README.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+should be a comment with text as `/ok-to-test` from any member to run
+the CI job.
+
+Here is a high-level schematic diagram of the CI pipeline for an
+end-to-end test:
+
+```
++--------+     +--------+     +--------+
+|        |     |        |     |        |
+|  root  +---->+  src   +---->+  bin   |
```
Comment (ldimaggi) [inline openshift-ci/README.md]: What is the "root" in this diagram? I can see how the source is built into bin from which an image is created and tests run - but what is the root?

File: openshift-ci/README.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+```
++--------+     +--------+     +----------------+
+|        |     |        |     |                |
+|  root  +---->+  src   +---->+ lint/unit test |
+|        |     |        |     |                |
++--------+     +--------+     +----------------+
+```
+
+
+All the steps mentioned here is taking place inside a work namespace.
```
Comment (ldimaggi) [inline openshift-ci/README.md]: Very small nit - "is" should be "are" - thanks!

File: openshift-ci/README.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+When you click on the job details from your pull request, you can see
+the name of the work namespace at top.  The name is going start with
+`ci-op-`.  The images created here will be available under this
+namespace.  The namespace can be accessed through
+`OPENSHIFT_BUILD_NAMESPACE` environment variable.
+
+## root
+
+As part of the CI pipeline, the first step is to create `root` image.
```
Comment (ldimaggi) [inline openshift-ci/README.md]: So, the root image contains the tools needed to build the Devconsole operator image?

File: openshift-ci/README.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+In fact, `root` is a tag created for the pipeline image.  This image
+contains all the tools including but not limited to Go compiler, git,
+kubectl, oc, and Operator SDK.
+
+The `root` image tag is created using this Dockerfile:
+`openshift-ci/Dockerfile.tools`
+
+## src
+
+This step clone the pull request branch with latest changes.  The
```
Comment (ldimaggi) [inline openshift-ci/README.md]: Small nit - Should be "This step clones"

File: openshift-ci/README.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+image.  As a result of this step an image named `src` is going to be
+created.
+
+In the CI configuration, there is a declaration like this:
+
+```
+canonical_go_repository: github.com/redhat-developer/devconsole-operator
+```
+
+The above line ensure the cloned source code goes into the specified
```
Comment (ldimaggi) [inline openshift-ci/README.md]: Small nit - should be: "The above line ensures"

File: openshift-ci/README.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+path: `$GOPATH/src/<canonical_go_repository>`.
+
+## bin
+
+This step runs the `build` Makefile target.  This step is taking place
+inside a container created from the `src` image created in the
+previous step.
+
+The `make build` produce operator binary image available under
```
Comment (ldimaggi) [inline openshift-ci/README.md]: Small nit - should be "produces an operator binary image"

File: openshift-ci/README.md [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+`devconsole-operator-registry` image build.  This intermediate image
+is going to be used as the base image for
+`devconsole-operator-registry`.  The original `operator-registry`
+image has a a different value for `OPENSHIFT_BUILD_NAMESPACE`
+environment variable, which must be set when that image was built.
+
+The Dockerfile used is available here:
+`openshift-ci/Dockerfile.registry.intermediate` As you can see, the
+Dockerfile file has only one line with a `FROM` statement, which is
+going to replaced with `operator-registry` image during image build.
```
Comment (ldimaggi) [inline openshift-ci/README.md]: Should be "going to be replaced"

File: openshift-ci/README.md [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+## tests
+
+### lint
+
+The lint runs the GolangCI Lint, YAML Lint and Operator Courier.
+GolangCI is a Go program, whereas the other two are written in Python.
+So, Python 3 is a perquisite to run lint.
+
+The GolangCI Lint program runs multiple Go lint tools against the
+repository.  GolangCI Lint is runs lint concurrently and complete
```
Comment (ldimaggi) [inline openshift-ci/README.md]: Should be:  "GolangCI Lint runs lint concurrently and completes execution in few seconds"

File: openshift-ci/README.md [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+`.yamllint`.
+
+The Operator Courier checks for the validity of Operator Lifecycle
+Manager (OLM) manifests.  That includes Cluster Service Version (CSV)
+files and CRDs.
+
+### test
+
+The `test` target runs the unit tests.  Some of the test make use mock
+objects. The unit tests doesn't require a dedicated OpenShift cluster
```
Comment (ldimaggi) [inline openshift-ci/README.md]: Should be:  test make use of mock objects. The unit tests don't

File: openshift-ci/README.md [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+requirement.  As of now there is no configuration provided to run
+GolangCI Lint.
+
+The YAML Lint tools validate all the YAML configuration files.  It
+excludes the `vendor` directory while running.  There is a
+configuration file at top-level directory of the source code:
+`.yamllint`.
+
+The Operator Courier checks for the validity of Operator Lifecycle
+Manager (OLM) manifests.  That includes Cluster Service Version (CSV)
```
Comment (ldimaggi) [inline openshift-ci/README.md]: It's good that the doc spells out the meaning of "CSV" - this should help new users.

File: openshift-ci/README.md [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+
+The `test` target runs the unit tests.  Some of the test make use mock
+objects. The unit tests doesn't require a dedicated OpenShift cluster
+unlike end-to-end tests.
+
+### e2e
+
+The `e2e` run an end-to-end test against an operator running inside
+the CI cluster pod but connected to a freshly created temporary
+Openshift 4 cluster.  It makes use the `--up-local` option provided by
```
Comment (ldimaggi) [inline openshift-ci/README.md]: Should be "makes use of"

File: openshift-ci/README.md [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+the Operator SDK testing tool.  It runs `test-e2e` Makefile target.
+
+### e2e-ci
+
+The `e2e-ci` target runs the end-to-end test against operator running
```
Comment (ldimaggi) [inline openshift-ci/README.md]: Should be "an operator"

File: openshift-ci/README.md [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+inside the freshly created OpenShift 4 cluster.  All the resources are
+created through scripts invoked from `test-e2e-ci` Makefile target.
+This Makefile target is designed to run exclusively on CI environment.
+
+### e2e-olm-ci
+
+The e2e-ci runs the end-to-end test against operator running inside
```
Comment (ldimaggi) [inline openshift-ci/README.md]: Should be "an operator"

File: openshift-ci/README.md (hunk 4 above)
Comment (baijum) [reply openshift-ci/README.md]: yes

File: openshift-ci/README.md (hunk 11 above)
Comment (baijum) [reply openshift-ci/README.md]: :+1:

File: openshift-ci/README.md (hunk 14 above)
Comment (baijum) [reply openshift-ci/README.md]: Don. ceba1627b1c5d4a8ced946a3b352b1416f4e6c3a

File: openshift-ci/README.md (hunk 13 above)
Comment (baijum) [reply openshift-ci/README.md]: Don. ceba1627b1c5d4a8ced946a3b352b1416f4e6c3a

File: openshift-ci/README.md (hunk 12 above)
Comment (baijum) [reply openshift-ci/README.md]: Don. ceba1627b1c5d4a8ced946a3b352b1416f4e6c3a

File: openshift-ci/README.md (hunk 10 above)
Comment (baijum) [reply openshift-ci/README.md]: Don. ceba1627b1c5d4a8ced946a3b352b1416f4e6c3a

File: openshift-ci/README.md (hunk 9 above)
Comment (baijum) [reply openshift-ci/README.md]: Don. ceba1627b1c5d4a8ced946a3b352b1416f4e6c3a

File: openshift-ci/README.md (hunk 8 above)
Comment (baijum) [reply openshift-ci/README.md]: Don. ceba1627b1c5d4a8ced946a3b352b1416f4e6c3a

File: openshift-ci/README.md (hunk 7 above)
Comment (baijum) [reply openshift-ci/README.md]: Don. ceba1627b1c5d4a8ced946a3b352b1416f4e6c3a

File: openshift-ci/README.md (hunk 3 above)
Comment (baijum) [reply openshift-ci/README.md]: Done. ceba1627b1c5d4a8ced946a3b352b1416f4e6c3a

File: openshift-ci/README.md (hunk 5 above)
Comment (baijum) [reply openshift-ci/README.md]: Done. ceba1627b1c5d4a8ced946a3b352b1416f4e6c3a

File: openshift-ci/README.md (hunk 6 above)
Comment (baijum) [reply openshift-ci/README.md]: Done. ceba1627b1c5d4a8ced946a3b352b1416f4e6c3a

File: openshift-ci/README.md (hunk 14 above)
Comment (rgarg1) [reply openshift-ci/README.md]: Pardon my ignorance, but are the e2e-ci tests a subset of e2e-olm-ci? What problems the former is expected to reveal that the latter cannot?

Comment (baijum) [reply openshift-ci/README.md]: The difference between `e2e-ci` and `e2e-olm-ci` is how the Kubernetes resources getting created. The resources include operator deployment, service account, role & role binding. The `e2e-ci` target uses a Makefile script (few `oc apply` commands) to create the resources whereas `e2e-olm-ci` uses Operator Lifecycle Manager (OLM) to create the resources.

Comment (rgarg1) [reply openshift-ci/README.md]: Thanks @baijum for helping me understand. Can you also answer the second part of my question?

Comment (baijum) [reply openshift-ci/README.md]: @rgarg1 Regarding revealing issues of the operator itself, there is no difference between these two end-to-end tests. The "e2e-olm-ci" verifies the Cluster Service Version file is correct, which is not happening in the previous "e2e-ci" test. However, the "e2e-olm-ci" setup is more complex, and there are more failure points. So having a simple setup to run end-to-end test would be helpful to iterate faster, even if there is some issue with OLM.

**Pull request merged**